src/services/prisma-datastore/CompositeAttributeStore.ts

```diff
@@ -0,0 +1,45 @@
+import { AttributeStore } from "../AttributeStore";
+
+// would be nice to generify this and implement as a generic proxy
+export default class CompositeAttributeStore implements AttributeStore {
+
+    constructor(private stores: AttributeStore[]) {}
+
+    private apply2(fn:(store: AttributeStore) => Promise<void>): Promise<void> {
+        if (this.stores.length==0) {
+            return Promise.resolve();
+        }
+
+        const tail = (p:Promise<void>, index: number): Promise<void> => {
+            if (index<this.stores.length) {
+                return p.then(() => {
+                    return tail(fn(this.stores[index]), index+1);
+                })
+            }
+            else {
+                return p;
+            }
+        };
+
+        return tail(fn(this.stores[0]), 1);
+    }
```

Comment:
`apply` looks better than `apply2`, we don't need `apply2` do we? I haven't thought about tail recursion in a while, good trip down memory lane.